toolkit/tools/pkg/imagecustomizerlib/extractpartitions.go

```diff
@@ -79,11 +118,165 @@ func compressWithZstd(partitionRawFilepath string) (partitionFilepath string, er
+
+	// Write the combined data back to the partition file.
+	err = os.WriteFile(partitionFilepath, newData, 0644)
 	if err != nil {
-		return "", fmt.Errorf("failed to remove raw file %s:\n%w", partitionRawFilepath, err)
+		return fmt.Errorf("failed to write skippable frame to partition file %s:\n%w", partitionFilepath, err)
 	}
 
-	return partitionRawFilepath + ".zst", nil
+	return nil
+}
+
+// Creates a skippable frame.
+func createSkippableFrame(magicNumber uint32, frameSize uint32, payload [4]uint32) (skippableFrame []byte) {
+	// Calculate the length of the byte array
+	lengthOfByteArray := 4 + 4 + (4 * len(payload))
+	// Define the Skippable frame
+	skippableFrame = make([]byte, lengthOfByteArray)
+	// Magic_Number
+	binary.LittleEndian.PutUint32(skippableFrame, magicNumber)
+	// Frame_Size
+	binary.LittleEndian.PutUint32(skippableFrame[4:8], frameSize)
+	// User_Data
+	binary.LittleEndian.PutUint32(skippableFrame[8:12], payload[0])
+	binary.LittleEndian.PutUint32(skippableFrame[12:16], payload[1])
+	binary.LittleEndian.PutUint32(skippableFrame[16:20], payload[2])
+	binary.LittleEndian.PutUint32(skippableFrame[20:24], payload[3])
+
+	logger.Log.Infof("Skippable frame has been created with the following metadata: %d", skippableFrame[8:24])
+
+	return skippableFrame
+}
+
+// Create user metadata that will be inserted into the skippable frame.
+func createSkippableFrameMetadata() (metadata [16]byte, err error) {
+	// Set the metadata to be a random 128-Bit number
+	metadata, err = generateRandom128BitNumber()
+	if err != nil {
+		return metadata, err
+	}
+	return metadata, nil
+}
+
+// Generates a Random 128-Bit number.
+func generateRandom128BitNumber() ([16]byte, error) {
+	var randomBytes [16]byte
+	_, err := rand.Read(randomBytes[:])
+	if err != nil {
+		return randomBytes, err
+	}
+	return randomBytes, nil
+}
+
+// Encodes a 128-Bit number into an array of uint32 in little-endian order.
+func Encode128BitLittleEndian(number [16]byte) [4]uint32 {
+	var encoded [4]uint32
+	for i := 0; i < 4; i++ {
+		offset := i * 4
+		encoded[i] = binary.LittleEndian.Uint32(number[offset : offset+4])
+	}
+	return encoded
+}
+
+// Calculates sha256sum hash for a given file
+func calculateSHA256(filePath string) (string, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	hash := sha256.New()
+	if _, err := io.Copy(hash, file); err != nil {
+		return "", err
+	}
+
+	hashInBytes := hash.Sum(nil)
+	hashString := hex.EncodeToString(hashInBytes)
+
+	return hashString, nil
+}
+
+// Decompress the .raw.zst partition file and verifies the hash matches with the source .raw file
+func verifySkippableFrameDecompression(rawPartitionFilepath string, rawZstPartitionFilepath string) (err error) {
+	// Decompressing .raw.zst file
+	decompressedPartitionFilePath := "build/decompressed.raw"
+	err = shell.ExecuteLive(true, "zstd", "-d", rawZstPartitionFilepath, "-o", decompressedPartitionFilePath)
```

Comment:
The files will potentially be pretty large. So, fully decompressing them will take a non-trivial amount of time. So, we probably shouldn't be doing this in the product code.

You should move this verification logic into UTs.

Reply:
Moved them into unit test file.